File: evaluator/012.cache-lifecycle/tests/test_report_cache.cc
#include <cmath>
#include <vector>

#include <gtest/gtest.h>

#include "inventory_report_service.h"
#include "summary_engine.h"

namespace {

class CountingSummaryEngine final : public nitr::case012::SummaryEngine {
 public:
  nitr::case012::InventorySummary Compute(
      const std::vector<nitr::case012::Product>& products) const override {
    compute_calls += 1;

    nitr::case012::InventorySummary summary;
    summary.product_count = static_cast<int>(products.size());
    for (const nitr::case012::Product& product : products) {
      if (product.units_in_stock <= product.reorder_threshold) {
        summary.low_stock_count += 1;
      }
      summary.inventory_value +=
          static_cast<double>(product.units_in_stock) * product.unit_price;
    }
    return summary;
  }

  mutable int compute_calls = 0;
};

bool NearlyEqual(double a, double b) {
  return std::fabs(a - b) < 1e-9;
}

TEST(Case012ReportCache, CachesRepeatedSummaryReads) {
  CountingSummaryEngine engine;
  nitr::case012::InventoryReportService service(&engine);

  service.ReplaceProducts({
      {"A-100", 10, 3, 2.5},
      {"B-200", 2, 5, 10.0},
  });

  const nitr::case012::InventorySummary first = service.GetSummary();
  const nitr::case012::InventorySummary second = service.GetSummary();

  EXPECT_EQ(engine.compute_calls, 1);
  EXPECT_EQ(first.product_count, 2);
  EXPECT_EQ(first.low_stock_count, 1);
  EXPECT_TRUE(NearlyEqual(first.inventory_value, 45.0));
  EXPECT_EQ(second.product_count, first.product_count);
  EXPECT_EQ(second.low_stock_count, first.low_stock_count);
  EXPECT_TRUE(NearlyEqual(second.inventory_value, first.inventory_value));
}

TEST(Case012ReportCache, ReplacingProductsInvalidatesCache) {
  CountingSummaryEngine engine;
  nitr::case012::InventoryReportService service(&engine);

  service.ReplaceProducts({
      {"A-100", 10, 3, 2.5},
      {"B-200", 2, 5, 10.0},
  });
  (void)service.GetSummary();

  service.ReplaceProducts({
      {"C-300", 1, 2, 7.0},
  });
  const nitr::case012::InventorySummary summary = service.GetSummary();

  EXPECT_EQ(engine.compute_calls, 2);
  EXPECT_EQ(summary.product_count, 1);
  EXPECT_EQ(summary.low_stock_count, 1);
  EXPECT_TRUE(NearlyEqual(summary.inventory_value, 7.0));
}

TEST(Case012ReportCache, UpsertingProductInvalidatesCache) {
  CountingSummaryEngine engine;
  nitr::case012::InventoryReportService service(&engine);

  service.ReplaceProducts({
      {"A-100", 10, 3, 2.5},
      {"B-200", 2, 5, 10.0},
  });
  (void)service.GetSummary();

  service.UpsertProduct({"B-200", 8, 5, 10.0});
  const nitr::case012::InventorySummary summary = service.GetSummary();

  EXPECT_EQ(engine.compute_calls, 2);
  EXPECT_EQ(summary.product_count, 2);
  EXPECT_EQ(summary.low_stock_count, 0);
  EXPECT_TRUE(NearlyEqual(summary.inventory_value, 105.0));
}

TEST(Case012ReportCache, ClearCacheForcesRecompute) {
  CountingSummaryEngine engine;
  nitr::case012::InventoryReportService service(&engine);

  service.ReplaceProducts({
      {"A-100", 10, 3, 2.5},
      {"B-200", 2, 5, 10.0},
  });
  (void)service.GetSummary();
  service.ClearCache();
  (void)service.GetSummary();

  EXPECT_EQ(engine.compute_calls, 2);
}

}  // namespace

File: evaluator/035.cache-lifecycle-python/tests/test_report_cache.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import sys
import unittest
from pathlib import Path

ARG_PARSER = argparse.ArgumentParser()
ARG_PARSER.add_argument("--case_root", required=True)
ARGS = ARG_PARSER.parse_args()

CASE_ROOT = Path(ARGS.case_root).resolve()
if str(CASE_ROOT) not in sys.path:
    sys.path.insert(0, str(CASE_ROOT))

from src.inventory_report_service import InventoryReportService
from src.product import Product
from src.summary_engine import InventorySummary, SummaryEngine


class CountingSummaryEngine(SummaryEngine):
    def __init__(self) -> None:
        self.compute_calls = 0

    def compute(self, products: list[Product]) -> InventorySummary:
        self.compute_calls += 1
        low_stock_count = sum(
            product.units_in_stock <= product.reorder_threshold for product in products
        )
        inventory_value = sum(
            product.units_in_stock * product.unit_price for product in products
        )
        return InventorySummary(len(products), low_stock_count, inventory_value)


def initial_products() -> list[Product]:
    return [
        Product("A-100", 10, 3, 2.5),
        Product("B-200", 2, 5, 10.0),
    ]


class ReportCacheTests(unittest.TestCase):
    def test_caches_repeated_summary_reads(self) -> None:
        engine = CountingSummaryEngine()
        service = InventoryReportService(engine)
        service.replace_products(initial_products())

        first = service.get_summary()
        second = service.get_summary()

        self.assertEqual(engine.compute_calls, 1)
        self.assertEqual(first, InventorySummary(2, 1, 45.0))
        self.assertEqual(second, first)

    def test_replacing_products_invalidates_cache(self) -> None:
        engine = CountingSummaryEngine()
        service = InventoryReportService(engine)
        service.replace_products(initial_products())
        service.get_summary()

        service.replace_products([Product("C-300", 1, 2, 7.0)])
        summary = service.get_summary()

        self.assertEqual(engine.compute_calls, 2)
        self.assertEqual(summary, InventorySummary(1, 1, 7.0))

    def test_updating_product_invalidates_cache(self) -> None:
        engine = CountingSummaryEngine()
        service = InventoryReportService(engine)
        service.replace_products(initial_products())
        service.get_summary()

        service.upsert_product(Product("B-200", 8, 5, 10.0))
        summary = service.get_summary()

        self.assertEqual(engine.compute_calls, 2)
        self.assertEqual(summary, InventorySummary(2, 0, 105.0))

    def test_inserting_product_invalidates_cache(self) -> None:
        engine = CountingSummaryEngine()
        service = InventoryReportService(engine)
        service.replace_products(initial_products())
        service.get_summary()

        service.upsert_product(Product("C-300", 1, 2, 7.0))
        summary = service.get_summary()

        self.assertEqual(engine.compute_calls, 2)
        self.assertEqual(summary, InventorySummary(3, 2, 52.0))

    def test_clear_cache_forces_recompute(self) -> None:
        engine = CountingSummaryEngine()
        service = InventoryReportService(engine)
        service.replace_products(initial_products())
        service.get_summary()

        service.clear_cache()
        service.get_summary()

        self.assertEqual(engine.compute_calls, 2)

    def test_cache_is_owned_per_service_instance(self) -> None:
        engine = CountingSummaryEngine()
        first_service = InventoryReportService(engine)
        second_service = InventoryReportService(engine)
        first_service.replace_products(initial_products())
        second_service.replace_products([Product("C-300", 1, 2, 7.0)])

        first_summary = first_service.get_summary()
        second_summary = second_service.get_summary()

        self.assertEqual(engine.compute_calls, 2)
        self.assertEqual(first_summary, InventorySummary(2, 1, 45.0))
        self.assertEqual(second_summary, InventorySummary(1, 1, 7.0))


if __name__ == "__main__":
    unittest.main(argv=[sys.argv[0]])
